qiskit/circuit/singleton.py

```diff
@@ -0,0 +1,250 @@
+
+
+def _impl_new(cls, *_args, **_kwargs):
+    # __new__ for the singleton instances.
+    raise TypeError(f"cannot create '{cls.__name__}' instances")
+
+
+def _impl_init_subclass(base, overrides):
+    # __init_subclass__ for the classes that make their children singletons (e.g. `SingletonGate`)
+
+    def __init_subclass__(cls, *, create_singleton=True, **kwargs):
+        super(base, cls).__init_subclass__(**kwargs)
+        if not create_singleton:
+            return
+
+        # We need to make a new type object that pokes in the overrides into the correct
+        # place in the method-resolution order.
+        singleton_class = _SingletonMeta.__new__(
+            _SingletonMeta,
+            f"_Singleton{cls.__name__}",
+            (cls, overrides),
+            # This is a dynamically generated class so it's got no module.  The slot layout of the
+            # singleton class needs to match any layout in the base.
+            {"__module__": None, "__slots__": (), "__new__": _impl_new, "_base_class": cls},
+            create_singleton=False,
+        )
+
+        # Make a mutable instance, fully instantiate all lazy properties, then freeze it.
+        cls._singleton_instance = cls(_force_mutable=True)
+        cls._singleton_instance._define()
+        cls._singleton_instance.__class__ = singleton_class
+
+    return classmethod(__init_subclass__)
+
+
+class _SingletonMeta(type(Instruction)):
+    # The inheritance above is to ensure metaclass compatibility with `Instruction`, though pylint
+    # doesn't understand that this is a subclass of `type`, so uses metaclass `self` conventions.
+
+    # pylint: disable=bad-classmethod-argument,no-self-argument
+
+    __slots__ = ()
+
+    def __new__(mcs, name, bases, namespace, *, overrides=None, **kwargs):
+        cls = super().__new__(mcs, name, bases, namespace, **kwargs)
+        if overrides is not None:
+            # The `__init_subclass__` logic is shared between `SingletonInstruction` and
+            # `SingletonGate`, but we can't do `__init_subclass__ = _impl_init_subclass(...)`
+            # inside the class definition bodies because it couldn't make `super()` calls.
+            cls.__init_subclass__ = _impl_init_subclass(cls, overrides)
+        return cls
+
+    def __call__(cls, *args, _force_mutable=False, **kwargs):
+        if not _force_mutable and not args and not kwargs:
+            # This class attribute is created by the singleton-creation base classes'
+            # `__init_subclass__` methods; see `_impl_init_subclass`.
+            return cls._singleton_instance
+        return super().__call__(*args, **kwargs)
+
+
+class _SingletonInstructionOverrides(Instruction):
+    """Overrides for all the mutable methods and properties of `Instruction` to make it
+    immutable."""
+
+    __slots__ = ()
+
+    def c_if(self, classical, val):
+        return self.to_mutable().c_if(classical, val)
```

Comment:
Oh, this is much neater and would have worked fine for the old class structure if we did `out = self.to_mutable();out.condition = (...)`. I really should have done a pass to clean up all the overrides after adding the `to_mutable()` method.

Reply:
Yeah - I think there's an instance in your `ControlledGate` PR where it could do a similar thing as well (or maybe it's a case that does `obj = obj.to_mutable(); obj.condition = ...` instead of `obj = obj.c_if(...)` - I forgot).